### PR TITLE
PHP 8.3 support added

### PR DESCRIPTION
**Pull request by glo71317**
|    Q          |   A
|-------------- | ------
| Documentation | yes/no
| Bugfix        | yes/no
| BC Break      | yes/no
| New Feature   | yes/no
| RFC           | yes/no
| QA            | yes/no

### CODE REVIEW

**Comment by glo71317** [comment]
Hi @Ocramius 

Few tests are failing due to some certification issue
**Error -** [Microsoft][ODBC Driver 18 for SQL Server]SSL Provider: [error:0A000086:SSL routines::certificate verify failed:self-signed certificate]

It seems some issue with CI/CD Any plan to fix this issue OR anything can we do?

**Comment by Ocramius** [comment]
Yep, the usual MSSQL mess.

@weierophinney can we perhaps drop testing MSSQL? It's such a massive waste of time and oxygen for other better RDBMS...

**Comment by glo17680** [comment]
@Ocramius / @weierophinney  is there any update on this. Can anyone please advise on the same ?

**Comment by Ocramius** [comment]
Either we rip out MSSQL testing (and mark it as untested somewhere in the docs), or we wait for Microsoft to release MSSQL support for PHP 8.3, which usually happens few weeks/months after release.

**Comment by weierophinney** [comment]
My feelings on the matter are:

- Not testing against MS SQL Server means we don't support it
- Not supporting it would be a BC break, requiring a new major
- We've marked this component as security-only, so a new major feels like a break with that

As such, my vote is we wait until MS has released a version of the extension for 8.3. If tests still fail at that time, we can regroup and figure out what to do from there; my guess is everything will "just work" as it has the last several releases.

**Comment by glo71317** [comment]
@Ocramius @weierophinney Any update on this?

**Comment by Ocramius** [comment]
No.

**Comment by driehle** [comment]
This is a bit unfortunate, as this blocks [laminas-log](https://github.com/laminas/laminas-log/pull/53), which is quite a central component in many applications.

How about extracting the namespace `Laminas\Db\Adapter\Driver\Sqlsrv` into a separate package? This would, obviously, require a new major release.

**Comment by weierophinney** [comment]
@driehle Considering this repo is security-only, we're not going to do a new major. It would likely make more sense to split the lamians-db related functionality in laminas-log into a new package.

I recognize folks are excited about the new PHP release but (a) it's been not even two weeks, and (b) not every organization is tracking PHP updates in order to release as soon as it drops. It's far better for us to wait and ensure that the component works as expected (which means running tests!) than it is to rush a release. Please be patient.

**Comment by glo71317** [comment]
@Ocramius Any update on this when we can expect laminas-db available with php 8.3 support

**Comment by Ocramius** [comment]
https://github.com/laminas/laminas-db/pull/293#issuecomment-1780945285

**Comment by rchrdkn** [comment]
Looks like this is the msphpsql issue that we're waiting on - https://github.com/microsoft/msphpsql/issues/1477

Let's hope they decide to release soon 🤞🏻

**Comment by panvid** [comment]
There is noting new?

**Comment by glo71317** [comment]
@Ocramius @weierophinney https://github.com/microsoft/msphpsql/releases/tag/v5.12.0 this released Can we prioritize this?

**Comment by glo71317** [comment]
I have fixed all the test failures and other deprecated dynamic properties issues, Only following error is there 

> Unable to connect laminasdb_test. Array
```
(
    [0] => Array
        (
            [0] => 08001
            [SQLSTATE] => 08001
            [1] => -1
            [code] => -1
            [2] => [Microsoft][ODBC Driver 18 for SQL Server]SSL Provider: [error:0A000086:SSL routines::certificate verify failed:self-signed certificate]
            [message] => [Microsoft][ODBC Driver 18 for SQL Server]SSL Provider: [error:0A000086:SSL routines::certificate verify failed:self-signed certificate]
        )
    [1] => Array
        (
            [0] => 08001
            [SQLSTATE] => 08001
            [1] => -1
            [code] => -1
            [2] => [Microsoft][ODBC Driver 18 for SQL Server]Client unable to establish connection. For solutions related to encryption errors, see https://go.microsoft.com/fwlink/?linkid=2226722
            [message] => [Microsoft][ODBC Driver 18 for SQL Server]Client unable to establish connection. For solutions related to encryption errors, see https://go.microsoft.com/fwlink/?linkid=2226722
        )
)
```

it seems it should be fixed from your end or suggest to us we can do anything for this.

**Comment by Ocramius** [comment]
SSounds like a new requirement by the MSSQL adapter? Can you try disabling SSL validation for MSSQL in the test suite?

**Comment by Xerkus** [comment]
This has to be done via connection options https://learn.microsoft.com/en-us/sql/connect/php/connection-options by setting `TrustServerCertificate` to 1

This is introduced by ODBC Driver 18

**Comment by glo71317** [comment]
> SSounds like a new requirement by the MSSQL adapter? Can you try disabling SSL validation for MSSQL in the test suite?

No idea how to disable ssl validation for MSSQL in the suite. so any idea?

**Comment by driehle** [comment]
https://github.com/laminas/laminas-db/blob/12385a47561c2d1feadf2bd114a67767541963a3/test/integration/Adapter/Platform/SqlServerTest.php#L36-L40

As Xerkus said, try adding `'TrustServerCertificate' => 1` there.

**Comment by glo71317** [comment]
> https://github.com/laminas/laminas-db/blob/12385a47561c2d1feadf2bd114a67767541963a3/test/integration/Adapter/Platform/SqlServerTest.php#L36-L40
> 
> As Xerkus said, try adding `'TrustServerCertificate' => 1` there.

I have made changes - https://github.com/laminas/laminas-db/pull/293/commits/2abd29d160bdca714e8198e786d4a1dcfcd0c4a4 but still having same issue

**Comment by Xerkus** [comment]
There is more than one place where `sqlsrv_connect()` is used. You need to find which of those need the option

File: .laminas-ci/install-sqlsrv-extension-via-pecl.sh
Comment (Xerkus) [inline .laminas-ci/install-sqlsrv-extension-via-pecl.sh:1]: Please drop this file entirely. sqlsrv is provided by CI container.

**Comment by glo71317** [comment]
> There is more than one place where `sqlsrv_connect()` is used. You need to find which of those need the option

I have updated all the places and fixed that issues, now lot tests are failing, Can you check these need to fixed still having some issue?

File: .laminas-ci.json
Comment (Xerkus) [inline .laminas-ci.json:13]: Please remove platform ignores, whole thing

**Comment by Xerkus** [comment]
Errors are caused by phpunit deprecations. Tests need to be migrated to newer phpunit version.

**Comment by Xerkus** [comment]
There is still one certificate validation error: 
```
 1) LaminasIntegrationTest\Db\Adapter\Platform\SqlServerTest::testQuoteValueWithSqlServer
PDOException: SQLSTATE[08001]: [Microsoft][ODBC Driver 18 for SQL Server]SSL Provider: [error:0A000086:SSL routines::certificate verify failed:self-signed certificate]

/github/workspace/test/integration/Adapter/Platform/SqlServerTest.php:54
```

**Comment by Xerkus** [comment]
> I have updated all the places and fixed that issues, now lot tests are failing, Can you check these need to fixed still having some issue?

Those are phpunit deprecation warnings. Tests need to be migrated to latest phpunit but i do not think they cause CI to fail. They would need to be dealt with in a separate PR.

**Comment by glo71317** [comment]
@Ocramius Can you review this PR changes and @Xerkus comments? what is your point of view on this?

File: src/Adapter/Driver/Sqlsrv/Connection.php
Comment (Xerkus) [inline src/Adapter/Driver/Sqlsrv/Connection.php]: This is a security feature that can not be applied as a default for adapter. Please revert.
This connection parameter needs to be properly configured via `'options'` from test case.

File: .laminas-ci.json
Comment (Xerkus) [inline .laminas-ci.json]: Remove platform ignores. There were needed for php release candidate and should not be present now.

File: .laminas-ci/install-sqlsrv-extension-via-pecl.sh
Comment (Xerkus) [inline .laminas-ci/install-sqlsrv-extension-via-pecl.sh:1]: Please remove this script. It is not needed.

sqlsrv extension is available in CI container directly. This script compiles extension from pecl and significantly increases CI run time.

File: phpcs.xml.dist
Comment (Xerkus) [inline phpcs.xml.dist]: Coding style violations need to be fixed, not ignored. Please revert.

File: test/integration/Adapter/Driver/Pdo/Postgresql/AdapterTest.php
Comment (Xerkus) [inline test/integration/Adapter/Driver/Pdo/Postgresql/AdapterTest.php]: I would prefer you to explicitly declare dynamically used properties rather than use this attriburte.
This annotation is hiding code quality issue and it is not suitable for this use case.

Same for all other uses of this attribute below.

**Review by Xerkus — CHANGES_REQUESTED** [review]
Maintenance work is frustrating but necessary. Thank you for your effort.

**Comment by Xerkus** [comment]
@glo71317 remaining error is because trust option is missing for pdo dsn:

```diff
--- a/test/integration/Adapter/Platform/SqlServerTest.php
+++ b/test/integration/Adapter/Platform/SqlServerTest.php
@@ -49,7 +49,7 @@ class SqlServerTest extends TestCase
             $this->adapters['pdo_sqlsrv'] = new PDO(
                 'sqlsrv:Server='
                     . getenv('TESTS_LAMINAS_DB_ADAPTER_DRIVER_SQLSRV_HOSTNAME')
-                    . ';Database=' . ($database ?: ''),
+                    . ';Database=' . ($database ?: '') .';TrustServerCertificate=1',
                 getenv('TESTS_LAMINAS_DB_ADAPTER_DRIVER_SQLSRV_USERNAME'),
                 getenv('TESTS_LAMINAS_DB_ADAPTER_DRIVER_SQLSRV_PASSWORD')
             );
```

File: phpcs.xml.dist
Comment (glo71317) [reply phpcs.xml.dist]: reverted

File: .laminas-ci/install-sqlsrv-extension-via-pecl.sh
Comment (glo71317) [reply .laminas-ci/install-sqlsrv-extension-via-pecl.sh:1]: done

File: .laminas-ci.json
Comment (glo71317) [reply .laminas-ci.json]: done

File: src/Adapter/Driver/Sqlsrv/Connection.php
Comment (glo71317) [reply src/Adapter/Driver/Sqlsrv/Connection.php]: reverted

File: .laminas-ci/install-sqlsrv-extension-via-pecl.sh
Comment (glo71317) [reply .laminas-ci/install-sqlsrv-extension-via-pecl.sh:1]: done

File: src/Adapter/Driver/IbmDb2/Connection.php
Comment (Xerkus) [inline src/Adapter/Driver/IbmDb2/Connection.php]: Again. Do not ignore coding standard issues. Fix them.

Comment (Xerkus) [reply src/Adapter/Driver/IbmDb2/Connection.php]: Since return value is expected
 
```suggestion
            return null;
```

Comment (glo71317) [reply src/Adapter/Driver/IbmDb2/Connection.php]: Are you want to remove empty retrun from here? error showing `Empty return statement not required here`

Comment (glo71317) [reply src/Adapter/Driver/IbmDb2/Connection.php]: done

**Comment by glo71317** [comment]
@Xerkus it seems tests are not running due to .laminas-ci/install-sqlsrv-extension-via-pecl.sh as you told to removed

File: src/Adapter/Driver/IbmDb2/Connection.php
Comment (Xerkus) [reply src/Adapter/Driver/IbmDb2/Connection.php]: There is a distinction between `return;` and `return null` or omitting return entirely.

When return types are declared both of those will result in a fatal error:

```
function (): void {
    return null;
};
```

```
function (): ?string {
    return;
};
```

In this case return type is not explicitly defined so php does not know which way is correct so there is no error.

We could technically omit return entirely for implicit return at the end.
When return value is not expected omitting `return;` is correct.
However, when return is expected but no return statement present we get a code quality issue since it is not entirely clear whether `return null` was implied or return statement was simply forgotten. Psalm normally would complain about this case.

**Comment by Xerkus** [comment]
> @Xerkus it seems tests are not running due to .laminas-ci/install-sqlsrv-extension-via-pecl.sh as you told to removed

Yes, it is referenced here: https://github.com/laminas/laminas-db/blob/12385a47561c2d1feadf2bd114a67767541963a3/.laminas-ci/pre-install.sh#L19

Please also remove that line.

File: README.md
Comment (Xerkus) [inline README.md]: This is not a correct. markdown lint complained about trailing whitespaces in `> `

**Comment by Xerkus** [comment]
I asked you to revert changes in adapter previously.  The correct way is to pass connection options as parameters for  `Connection` instance: `new Connection([..., 'options' => ['TrustServerCertificate' => '1']]);`

I looked at the tests code and best place seems to be here: 
```diff
--- a/test/unit/Adapter/Driver/Sqlsrv/AbstractIntegrationTest.php
+++ b/test/unit/Adapter/Driver/Sqlsrv/AbstractIntegrationTest.php
@@ -38,6 +38,8 @@ abstract class AbstractIntegrationTest extends TestCase
             $this->variables[$name] = getenv($value);
         }
 
+        $this->variables['options'] = ['TrustServerCertificate' => '1'];
+
         if (! extension_loaded('sqlsrv')) {
             $this->fail('The phpunit group integration-sqlsrv was enabled, but the extension is not loaded.');
         }
```

**Comment by Xerkus** [comment]
Also you seem to have missed re-adding changes for some ` sqlsrv_connect()` calls in tests.

**Comment by glo71317** [comment]
> sqlsrv_connect

I don't think as i can see here i removed `src/Adapter/Driver/Sqlsrv/Connection.php` as u suggested

**Comment by Xerkus** [comment]
> > sqlsrv_connect
> 
> I don't think as i can see here i removed `src/Adapter/Driver/Sqlsrv/Connection.php` as u suggested

Nevermind, looking at a wrong commit.

File: test/integration/Adapter/Platform/SqlServerTest.php
Comment (Xerkus) [inline test/integration/Adapter/Platform/SqlServerTest.php]: AFAIK spaces in DSN are invalid.

Comment (glo71317) [reply test/integration/Adapter/Platform/SqlServerTest.php]: If i removed than phpcs failed

**Comment by glo71317** [comment]
All issues resolved from my end now  and pr is green now thank you so much for your quick support and review

**Comment by Xerkus** [comment]
You are missing signoff for your commits. See DCO check for details

**Comment by glo71317** [comment]
> You are missing signoff for your commits. See DCO check for details

Shall i use following command
`git rebase HEAD~8 --signoff`
`git push --force-with-lease origin php8.3_support `

File: test/integration/Adapter/Platform/SqlServerTest.php
Comment (Xerkus) [inline test/integration/Adapter/Platform/SqlServerTest.php]:
```suggestion
                    . ';Database=' . ($database ?: '') . ';TrustServerCertificate=1',
```

Comment (glo71317) [reply test/integration/Adapter/Platform/SqlServerTest.php]: done

**Review by Xerkus — APPROVED** [review]

**Pull request merged**
